### PR TITLE
Change resolved banks on pbl payments

### PR DESCRIPTION
**Pull request by arti0090**
Left TODO:

- [x] Check the payment methods to see if they are enabled at the moment (for valid channel list)
- [x] Payments different than banks do not work on vendor level (attached screenshot), maybe it is sandbox api issue?
Zrzut ekranu 2024-10-30 o 12 33 46

note to 2nd todo: partialy works, some instalments do work some not but it is coming from tpay API not from plugin

### CODE REVIEW

**Review by lchrusciel — DISMISSED** [review]

**Review by lchrusciel — APPROVED** [review]

File: src/Repository/FindByChannelWithGatewayConfigTrait.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+trait FindByChannelWithGatewayConfigTrait
+{
+    public function findByChannelAndGatewayConfigNameWithGatewayConfig(ChannelInterface $channel, string $gatewayConfigName): array
+    {
+        return $this->createQueryBuilder('o')
+            ->addSelect('gatewayConfig')
+            ->leftJoin('o.gatewayConfig', 'gatewayConfig')
+            ->where('o.enabled = :enabled')
+            ->andWhere(':channel MEMBER OF pm.channels')
```
Comment (lchrusciel) [inline src/Repository/FindByChannelWithGatewayConfigTrait.php:21]:
```suggestion
        return $this->createQueryBuilder('o')
            ->addSelect('gatewayConfig')
            ->leftJoin('o.gatewayConfig', 'gatewayConfig')
            ->where('o.enabled = :enabled')
            ->andWhere(':channel MEMBER OF o.channels')
```

File: src/Model/GatewayConfigInterface.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
+
+namespace CommerceWeavers\SyliusTpayPlugin\Model;
+
+use Payum\Core\Model\GatewayConfigInterface as BaseGatewayConfigInterface;
+use Payum\Core\Security\CryptedInterface;
+
+interface GatewayConfigInterface extends BaseGatewayConfigInterface, CryptedInterface
```
Comment (coldic3) [inline src/Model/GatewayConfigInterface.php:10]: Is this needed? I can't find an implementation of this interface.

File: src/Repository/FindByChannelWithGatewayConfigTrait.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * @mixin EntityRepository
+ */
+trait FindByChannelWithGatewayConfigTrait
+{
+    public function findByChannelAndGatewayConfigNameWithGatewayConfig(ChannelInterface $channel, string $gatewayConfigName): array
+    {
+        return $this->createQueryBuilder('o')
+            ->addSelect('gatewayConfig')
+            ->leftJoin('o.gatewayConfig', 'gatewayConfig')
```
Comment (coldic3) [inline src/Repository/FindByChannelWithGatewayConfigTrait.php:19]:
```suggestion
            ->innerJoin('o.gatewayConfig', 'gatewayConfig', Join::WITH, 'gatewayConfig.gatewayName = :gatewayName')
```

and then we can remove the line `->andWhere('gatewayConfig.gatewayName = :gatewayName')`

Comment (coldic3) [reply src/Repository/FindByChannelWithGatewayConfigTrait.php:19]: or you can leave where statement as it is but I'd still replace a left join with an inner join.

File: src/Tpay/Provider/ValidTpayChannelListProvider.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            $tpayGatewayConfig = $paymentMethod->getGatewayConfig();
+
+            if (null === $tpayGatewayConfig) {
+                continue;
+            }
+
+            $tpayGatewayConfig->decrypt($this->cypher);
+            $config = $tpayGatewayConfig->getConfig();
+
+            if (!array_key_exists('type', $config)) {
```
Comment (coldic3) [inline src/Tpay/Provider/ValidTpayChannelListProvider.php:52]: This is not unit tested

File: src/Model/GatewayConfigInterface.php (hunk 2 above)
Comment (lchrusciel) [reply src/Model/GatewayConfigInterface.php]: We have problem of not having CryptedInterface on the object within Sylius. But I've managed with workaround

File: src/Tpay/Provider/ValidTpayChannelListProvider.php (hunk 4 above)
Comment (coldic3) [reply src/Tpay/Provider/ValidTpayChannelListProvider.php:53]: In this test, there is still type provided https://github.com/CommerceWeavers/SyliusTpayPlugin/pull/140/files#diff-7a1d71f37f8fe786106bc96fef164a5fd39f598e88c9c38f44062c760d4e4f3aR214

I think we should test the case where the `$config` array is empty for example.

Comment (coldic3) [reply src/Tpay/Provider/ValidTpayChannelListProvider.php:53]: Oh ok, missed that https://github.com/CommerceWeavers/SyliusTpayPlugin/pull/140/files#diff-7a1d71f37f8fe786106bc96fef164a5fd39f598e88c9c38f44062c760d4e4f3aR223

**Pull request merged**